packages/inertia/src/inertia.js

```diff
@@ -63,42 +64,80 @@ export default {
       region.scrollTop = 0
       region.scrollLeft = 0
     })
+    this.saveScrollPositions()
+
+    if (window.location.hash) {
+      const el = document.getElementById(window.location.hash.slice(1))
+      if (el) {
+        el.scrollIntoView()
+      }
+    }
   },
 
-  restoreScrollPositions(page) {
-    if (page.scrollRegions) {
+  restoreScrollPositions() {
+    if (this.page.scrollRegions) {
       this.scrollRegions().forEach((region, index) => {
-        region.scrollTop = page.scrollRegions[index].top
-        region.scrollLeft = page.scrollRegions[index].left
+        region.scrollTop = this.page.scrollRegions[index].top
+        region.scrollLeft = this.page.scrollRegions[index].left
       })
     }
   },
 
-  navigationType() {
-    if (window.performance && window.performance.getEntriesByType('navigation').length) {
-      return window.performance.getEntriesByType('navigation')[0].type
+  isBackForwardVisit() {
+    return window.history.state
+      && window.performance
+      && window.performance.getEntriesByType('navigation').length
+      && window.performance.getEntriesByType('navigation')[0].type === 'back_forward'
+  },
+
+  handleBackForwardVisit() {
+    this.setPage(window.history.state, { preserveScroll: true }).then(() => {
+      this.restoreScrollPositions()
+    })
+  },
+
+  locationVisit(url, preserveScroll) {
+    try {
+      window.sessionStorage.setItem('inertiaLocationVisit', JSON.stringify({ preserveScroll }))
+      window.location.href = url
+    } catch (error) {
+      return false
     }
   },
 
-  isInertiaResponse(response) {
-    return response && response.headers['x-inertia']
+  isLocationVisit() {
+    try {
+      return window.sessionStorage.getItem('inertiaLocationVisit') !== null
+    } catch (error) {
+      return false
+    }
   },
 
-  isHardVisit(response) {
+  handleLocationVisit(page) {
+    const locationVisit = JSON.parse(window.sessionStorage.getItem('inertiaLocationVisit'))
+    window.sessionStorage.removeItem('inertiaLocationVisit')
+    page.url += window.location.hash
+    page.rememberedState = window.history.state ? window.history.state.rememberedState : {}
+    page.scrollRegions = window.history.state ? window.history.state.scrollRegions : []
```

Comment:
Made you some syntax suggestions which simplify code at several part. You take what you want. 🙂

```suggestion
    page.rememberedState = window.history.state?.rememberedState ?? {}
    page.scrollRegions = window.history.state?.scrollRegions ?? []
```